File: kangaroo_app_sdk_ios/ios/Classes/features/userAmazonCatalogue/UserAmazonCatalogueHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class UserAmazonCatalogueHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/get_user_amazon_catalogue"

    var eventChannel: String = "customer_sdk/events/get_user_amazon_catalogue"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await UserAmazonCatalogueHandler.getUserAmazonCatalogue(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func getUserAmazonCatalogue(call: FlutterMethodCall) async -> String? {


        


        

        guard let args = call.arguments else {
            return nil
        }
        do {
       if let myArgs = args as? [String: Any] {
          let overrideHeaders = myArgs["overrideHeaders"] as? [String: String]
                        guard let pageNumber = myArgs["pageNumber"] as? Int32 else {return nil}
                guard let perPage = myArgs["perPage"] as? Int32 else {return nil}
                guard let keywords = myArgs["keywords"] as? String? else {return nil}
                guard let filters = myArgs["filters"] as? String? else {return nil}

        let result = try await UserAmazonCatalogueApi().getUserAmazonCatalogue(
                overrideHeaders: overrideHeaders,
                pageNumber: pageNumber,
                perPage: perPage,
                keywords: keywords,
                filters: filters
           ).serializeUserAmazonCatalogueApiResult()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
                }
            }
        }
        catch {
            return nil
        }
        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        UserAmazonCatalogueApi().userAmazonCatalogueState
            .serializeUserAmazonCatalogueState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}

File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/userAmazonCatalogue/UserAmazonCatalogueHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.userAmazonCatalogue

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.userAmazonCatalogue.UserAmazonCatalogueApi
import features.userAmazonCatalogue.models.UserRewardsModel

import features.userAmazonCatalogue.serializeUserAmazonCatalogueState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class UserAmazonCatalogueHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/get_user_amazon_catalogue"

    override val eventChannel: String
        get() = "customer_sdk/events/get_user_amazon_catalogue"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return getUserAmazonCatalogue(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun getUserAmazonCatalogue(call: MethodCall): String? {
            val result = UserAmazonCatalogueApi().getUserAmazonCatalogue(
                overrideHeaders = call.argument<Map<String, String>>("overrideHeaders") as Map<String, String>?,
                pageNumber = call.argument<Int>("pageNumber") as Int,
                perPage = call.argument<Int>("perPage") as Int,
                keywords = call.argument<String?>("keywords") as String?,
                filters = call.argument<String?>("filters") as String?
            ).toJsonResult<UserRewardsModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        UserAmazonCatalogueApi().userAmazonCatalogueState.serializeUserAmazonCatalogueState().observe { result ->
            pushSerializedResultToEventSink<UserAmazonCatalogueApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}
